### PR TITLE
Fix get_index_in_direction for regular meshes

### PR DESCRIPTION
**Pull request by JoffreyDorville**
# Description

For regular meshes, the function `get_index_in_direction` is returning 0 instead of 1 when the tested coordinate is coincident with the lower boundary. This results in points being incorrectly seen outside the mesh when they lie on surfaces corresponding to the lower boundaries of the mesh.

For rectilinear meshes, the function `get_index_in_direction` works as expected.

I added a cpp unit test to illustrate the problem and to show that this PR fixes it.

This fix also corrected the results of the `weightwindows/survival_biasing` regression tests which have been updated. I cross-checked that using an equivalent rectilinear mesh instead of the regular mesh in tests/regression_tests/weightwindows/survival_biasing/test.py produces the same results as the updated ones with the fix.

Rectilinear mesh used:

```
mesh = openmc.RectilinearMesh()
mesh.x_grid = np.linspace(0.0, 160.0, 21)
mesh.y_grid = np.linspace(0.0, 160.0, 21)
mesh.z_grid = np.linspace(0.0, 160.0, 2)
```

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 18) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: src/mesh.cpp [hunk 1]
```diff
@@ -1472,7 +1472,11 @@ RegularMesh::RegularMesh(hid_t group) : StructuredMesh {group}
 
 int RegularMesh::get_index_in_direction(double r, int i) const
 {
-  return std::ceil((r - lower_left_[i]) / width_[i]);
+  if (r == lower_left_[i]) {
```
Comment (GuySten) [inline src/mesh.cpp]: I think this can be simplified to floor + 1.

Comment (JoffreyDorville) [reply src/mesh.cpp]: Will it not put the problem on the upper boundary instead?

Comment (GuySten) [reply src/mesh.cpp]: I do not think so.
I think floor + 1 is equivalent to your current implementation.
I've tried it and the cpp tests you provided are passing.
The floor + 1 solution also resembles the implementation in rectilinear mesh.

Comment (JoffreyDorville) [reply src/mesh.cpp]: The cpp tests are just testing the lower boundary case as the upper boundary case was working correctly. So you won't see the problem with floor + 1 with the tests as they are written.

You can add the following lines to the tests to also cover the upper boundary check:

For TEST_CASE("Test get_index_in_direction - regular mesh"):

```
REQUIRE(mesh.get_index_in_direction(1.99, 0) == 3);
REQUIRE(mesh.get_index_in_direction(2.0, 0) == 3);
```

For TEST_CASE("Test get_index_in_direction - rectilinear mesh"):

```
REQUIRE(mesh.get_index_in_direction(9.99, 0) == 3);
REQUIRE(mesh.get_index_in_direction(10.0, 0) == 3);
```

Comment (JoffreyDorville) [reply src/mesh.cpp]: I just added these checks in the cpp tests so that all boundaries are checked to avoid any confusion.

Comment (GuySten) [reply src/mesh.cpp]: You're right my "simplification" makes the upper boundary case wrong.
It's nice to have the upper boundary cases in the cpp test suite.

**Review by GuySten — APPROVED** [review]
Looks good to me.

**Review by paulromano — CHANGES_REQUESTED** [review]
I'm not sure I agree with the change. When the position being tested is on the boundary between two mesh elements, it is ambiguous as to whether it should be considered as being in the mesh element to the left or to the right. To me it seems more consistent to always pick one direction.

We have other ways of handling this ambiguity elsewhere in the code such as 1) pushing the position forward by `TINY_BIT` to move it off the boundary, or 2) Looking at the dot product of `u` and the surface normal to decide which one to choose.

**Comment by GuySten** [comment]
I think using the particle direction is a good solution.
In any case, equivalent regular and rectilinear meshes should have the same behavior.
Currently they behave differently.

**Comment by JoffreyDorville** [comment]
Thanks for the feedback @paulromano!

I just added the direction check for regular and rectilinear meshes. The solution is even simpler than pushing the particle or using the surface normal.

For cylindrical and spherical meshes, the task is a bit more involved because of specific cases (i.e., non-convex shapes). I have some ideas on how to handle these cases but it can be on a different PR as it is not required for the current state of the temperature field PR.

I have not looked into unstructured meshes for the moment.

I changed the signature of several functions that needed the new direction parameter, which also had the drawback of changing the signature of functions that did not need it. Let me know what you prefer for the Mesh interface.

File: src/mesh.cpp [hunk 2]
```diff
@@ -1470,9 +1471,18 @@ RegularMesh::RegularMesh(hid_t group) : StructuredMesh {group}
...
 {
-  return std::ceil((r - lower_left_[i]) / width_[i]);
+  int idx = std::ceil((r - lower_left_[i]) / width_[i]);
+
+  // If on upper boundary with positive direction, use next index
+  if (r == lower_left_[i] + width_[i] * idx) {
+    if (u > 0.0) {
+      idx++;
+    }
+  }
```
Comment (paulromano) [inline src/mesh.cpp]: This doesn't really match mentally what I was expecting. I would have thought that a particle that is coincident with any mesh boundary would account for the direction, not just the upper boundary. Additionally, the check for the upper boundary is done using floating-point equality, which seems like it may fail even when the particle is exactly on the boundary because of rounding/truncation.

File: src/mesh.cpp [hunk 3]
```diff
@@ -1711,9 +1721,25 @@ int RectilinearMesh::set_grid()
...
 }
 
-int RectilinearMesh::get_index_in_direction(double r, int i) const
+int RectilinearMesh::get_index_in_direction(double r, double u, int i) const
 {
-  return lower_bound_index(grid_[i].begin(), grid_[i].end(), r) + 1;
+  int idx = lower_bound_index(grid_[i].begin(), grid_[i].end(), r) + 1;
+
+  // If on lower boundary with negative direction, use previous index
+  if (r == grid_[i][idx - 1]) {
```
Comment (paulromano) [inline src/mesh.cpp]: If `r` is below the first grid point, we would have that idx == 0, which means that `grid_[i][idx - 1]` would be out of bounds.

File: src/mesh.cpp [hunk 4]
```diff
@@ -1711,9 +1721,25 @@ int RectilinearMesh::set_grid()
...
+    if (u < 0) {
+      idx--;
+    }
+  }
+
+  // If on upper boundary with positive direction, use next index
+  if (r == grid_[i][idx]) {
```
Comment (paulromano) [inline src/mesh.cpp]: Same consideration here: if `r` is above the last grid point, this will be out of bounds.

**Review by paulromano — CHANGES_REQUESTED** [review]
I made some comments below regarding the current implementation. However, I think the bigger issue is that this seems like an overly complicated solution for a relatively minor consistency issue. Sorry to go back and forth here, but I would prefer either:
1. Reverting to the original fix, `if (r == lower_left_[i]) return 1`, or
2. Change the implementation of `RectilinearMesh::get_index_in_direction` to use `std::lower_bound` directly (`std::lower_bound(grid_[i].begin(), grid_[i].end(), r) - grid_[i].begin()`). The reason it currently behaves differently than RegularMesh is because `lower_bound_index` has a special case for the first value in an array.

I lean toward solution 2 because it would treat all mesh cells equivalently with respect to behavior on the boundary (each mesh cells owns its upper boundary but not its lower).

File: src/mesh.cpp (hunk 2 above)
Comment (JoffreyDorville) [reply src/mesh.cpp]: I agree that the floating-point equality is definitely not ideal.

This only needs to be checked against upper boundary because `idx` is given by `std::ceil`. This means that if you are right on a boundary, `std::ceil` will give you the index of the previous cell, making this boundary the upper boundary of the identified cell by default.

File: src/mesh.cpp (hunk 3 above)
Comment (JoffreyDorville) [reply src/mesh.cpp]: Good catch! This should be protected so that `idx` is only contained between `1` and `n_cells`. I won't modify it though as we might change strategy here.

File: src/mesh.cpp (hunk 4 above)
Comment (JoffreyDorville) [reply src/mesh.cpp]: Same comment as above:

> Good catch! This should be protected so that `idx` is only contained between `1` and `n_cells`. I won't modify it though as we might change strategy here.

**Comment by JoffreyDorville** [comment]
I agree that the current modifications are relatively heavy for such a simple problem!

The initial objective of my correction is to get a point seen as inside a `RegularMesh` if it is either on a upper or lower boundary. While I agree that solution 2 brings more consistency in a sense, solution 1 is an easy reach to get to my objective. With solution 2, I would need to add an extra check in the temperature field implementation. In any case, both options work as long as we document this choice so that we can continue to be consistent in the future.

**Comment by JoffreyDorville** [comment]
Going back to it, I see that there is a fundamental difference in the way `get_index_in_direction` works between `RegularMesh` and `RectilinearMesh`. For `RectilinearMesh`, the returned index is bound between 0 and n_cell + 1. For `RegularMesh`, it can be any value (as if we were looking at an infinite grid). `test_mesh`.cpp shows several examples of this.

@paulromano  I can look at it in more details but is there an obvious reason for the `RegularMesh` implementation to allow such behavior?

Given this difference, I understand why the asymmetry (cells own their upper boundary, but the first cell also owns its lower boundary) makes sense if the index is bound in the case of `RectilinearMesh`. It also facilitates the test to know if you are inside the mesh or not. However, if you can access all cells in an infinite grid, like in the `RegularMesh` case, the asymmetry feels more inconsistent.

So I see 2 options here:
1) Bound the returned index for `RegularMesh` just like what is done for `RectilinearMesh` and then apply `if (r == lower_left_[i]) return 1`,
2) Change the implementation of all other `get_index_in_direction` functions so that we remove the asymmetry. We would also need to provide an explicit function to test that we are inside the mesh (check index + check if we are on the lower boundary) for all mesh classes.

Option 1 seems easier to implement if the infinite grid feature is not necessary, but that keeps the implicit rule of a different behavior for the first cell. Option 2 is cleaner but requires a bit more work and verification (I want to make sure that I understand the impact of removing the asymmetry, particularly in `raytrace_mesh`).

I think I will probably go with option 2 and see where we land!

**Review by paulromano — APPROVED** [review]
@JoffreyDorville For simplicity, I updated your branch with option 1 so that at least the two mesh types are consistent. We could revisit the first vs last inconsistency another time but at the end of the day, it is an arbitrary choice.

**Comment by JoffreyDorville** [comment]
I agree this is probably the easiest way to proceed.
The option 2 version I started to implement was way too complicated.
Thanks for merging this!
